app/controllers/application_controller.rb

```diff
@@ -5,13 +5,33 @@ class ApplicationController < ActionController::Base
 
   before_action :configure_permitted_parameters, if: :devise_controller?
   before_action :authenticate_user!
+  before_action :set_locale
 
   rescue_from ActiveRecord::RecordNotFound, with: :not_found
 
-  breadcrumb I18n.t('home'), :root_path
+  breadcrumb I18n.t('home_nav'), :root_path
+
+  def set_locale
+    language = language_params
+
+    if user_signed_in?
+      current_user.update(language: language) if language
+      I18n.locale = current_user.language
```

Comment:
bu noktada `current_user.language`'in nil olmadığına emin miyiz?  bu kodda sanki son söz `I18n.locale = ???  || I18n.default_locale` olmalı.  şöyle bir şey?

```ruby
def set_locale
  language = language_params

  I18n.locale =
    if user_signed_in?
      current_user.update(language: language) if language
      current_user.language
    else
      language
    end || I18n.default_locale
end
```

Reply:
(bunun için şimdilik bir şey yapmak gerekmiyor ama) locale ve language arasındaki semantik farklılığa da dikkat edelim.  locale dil tercihi dışındaki ayarları da kapsıyor (ör. tercih edilen ölçüm sistem, zaman gösterimi vs)

Reply:
Nil olmadığına eminiz hocam çünkü db/migrate şu şekilde:

```
add_column :users, :language, :string, default: 'tr'
```

kullanıcı oluştuğunda profil dili default olarak 'tr'.

Reply:
Dili 'tr' set ederse, sadece çeviriler değil, zaman birimleri - ölçü birimleri gibi şeyleri de o dilde görüyor - bu yüzden locale dedim.

Reply:
Ek olarak uygulama dilini set etmenin başka bir yolu yok.

`I18n.default_locale` ve `I18n.locale` benim seçtiğim metod isimleri değil, I18n'de böyle. I18n.language gibi bir şey set edemiyoruz Rails'te. `locale` set edildikten sonra tercüme gereken yerde t(), localization gereken yerde (tarih-saat vs.) l() devreye giriyor.

Reply:
I18n API'nin farkındayım.  sanırım yaptığın şu kısım http://guides.rubyonrails.org/i18n.html#managing-the-locale-across-requests

beni (bir an için) ilgilendiren tarafı request'ler arası değil de (veya ona ilave olarak) kullanıcı tercihi olsaydı buna nasıl bir isim verilebileceği idi (http://guides.rubyonrails.org/i18n.html#setting-the-locale-from-user-preferences).  rails dokümantasyonu görebildiğim kadarıyla her yerde buna language değil locale demiş.  ama öte yandan kimliklendirme açısından bakarsak ilgilendiğimiz şemalarda (inetOrgPerson, eduPerson, SHAC) locale şeklinde bir attribute görmüyorum.  örneğin inetOrg ve eduPerson'da `preferredLanguage`, SHAC'ta ise `schacMotherTongue` var en yakın.  o taraftan bakarsak language daha uyuyor.  dediğim gibi not ediyorum sadece...

Reply:
bizde şuan User modelinde 'language' olarak tutuluyor hocam kullanıcı tercihi. default olarak 'tr' geliyor ama öğrenci İngilizce'ye set ederse 'en' olarak değişiyor.

İsterseniz o field'ın ismini prefered_language gibi bir şeyle (veya kimliklendirmeye uygun başka bir isimle) değiştirebilirim.

Reply:
bence bu kararı defer edelim.  review yaparken sesli düşünüyorum :-)  hele bir user iyi kötü netleşsin, toplu halde karar veririz.  ve evet, `preferred_language` çok iyi bir seçim.